### PR TITLE
Speed up TaskAlignedAssigner on CPU and MPS by removing size-2 and middle-axis reductions

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

Two commits in `ultralytics/utils/tal.py`, one per reduction shape, each reviewable on its own.

- `RotatedTaskAlignedAssigner.select_candidates_in_gts` built `(b, n_boxes, h*w, 2)` temporaries and reduced them over a size-2 axis with `sum(dim=-1)`; the two dot products are now written per coordinate on `(b, n_boxes, h*w)` tensors, the indexing the base-class method already uses (commit 1). Low-precision inputs round identically in both forms: the old product tensor was already stored in the input dtype, so its float-accumulated `sum` only ever added two rounded values and rounded once, exactly what the per-coordinate add does (0 differing on 100k random fp16 pairs, 0 flipped of 983,040 near-edge pairs at homogeneous fp16/bf16 on CPU and MPS); every training path hands the assigner float32 boxes, and under AMP the fp16/bf16 anchors promote to float32 at the first subtraction.
- `TaskAlignedAssigner.select_highest_overlaps` ran two `argmax` over the middle axis of `(b, n_max_boxes, h*w)`, a per-output loop on CPU; both become `max(dim)`, and because every anchor serves at most one gt after the dedup, the final `max` also yields `fg_mask` and the two later `sum(-2)` lines are deleted (commit 2). `fg_mask` is returned in the mask's dtype; its only readers compare it `> 0` and cast it `.bool()`.
- outputs unchanged: `torch.equal` on 343 cases (commit 1) and 5808 cases per device (commit 2) including captured real inputs, fp16/bf16/fp64, forced ties, NaN and inf rows, kernel-boundary sizes, non-contiguous views, topk2-removed assignments, plus deterministic dota128 and coco128 training runs with identical per-iteration losses and final weights; the same fuzz on CUDA (Tesla T4, torch 2.10+cu128) returns 0 mismatches, including 0 flipped of 983,040 near-edge anchors at homogeneous fp16/bf16 and 0 differing of 200k pairs for the size-2 `sum` against the pointwise add, so the low-precision rounding is identical there as well

## measured

Apple M4, torch 2.14, and a Kaggle Tesla T4, torch 2.10+cu128: live methods on captured real inputs (yolo26n-obb on dota128 at imgsz 1024 batch 8; yolo26n on coco128 at imgsz 640 batch 16) and the full `E2ELoss` criterion on seeded mosaic batches, ranges over 3 (methods) or 2 (criterion) ABBA blocks, ms.

|                                                                             | before                        | after                         |
| --------------------------------------------------------------------------- | ----------------------------- | ----------------------------- |
| `select_candidates_in_gts`, 12 captured OBB calls, CPU 8 thr / 1 thr / MPS  | 282–308 / 1248–1300 / 236–242 | 92–95 / 181–187 / 189–191     |
| `select_highest_overlaps`, 6 captured detect calls, CPU 8 thr / 1 thr / MPS | 208–211 / 201–203 / 21–26     | 18.4–19.0 / 46.8–47.4 / 21–24 |
| detect criterion per call, CPU, three batches                               | 144–177 / 139–173 / 212–313   | 79–91 / 83–102 / 136–169      |
| detect criterion per call, MPS, three batches                               | 77–88 / 105–144 / 169–216     | 78–85 / 98–122 / 174–203      |
| OBB criterion per call, CPU, three batches, both commits                    | 158–197 / 264–272 / 365–385   | 48–67 / 105 / 179–227         |
| `select_candidates_in_gts`, CUDA T4, b8 × 21504 anchors, n_boxes 14 / 27 / 47 / 54 / 108 | 1.63 / 2.58 / 4.00 / 4.51 / 8.44 | 1.56 / 2.41 / 3.73 / 4.19 / 7.73 |
| `select_highest_overlaps`, CUDA T4, (16, n, 8400), n 18 / 30 / 35 / 46, topk 10 and 7/1 | 0.83–3.42 | 0.63–2.45 (1.32–1.40x) |
| detect criterion per call, CUDA T4, three batches | 33.5 / 25.1 / 32.3 | 31.1 / 24.4 / 32.0 |
| OBB criterion per call, CUDA T4, three batches, both commits | 24.1 / 28.8 / 35.5 | 22.2 / 27.8 / 33.8 |

The criterion runs each method twice per training step. Loss and gradient fingerprints are identical in every arm on CPU and MPS; on CUDA the loss fingerprints agree between arms wherever cuDNN reproduces itself across processes (deterministic mode was not pinned in that run).

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Optimizes task assignment by replacing small and middle-axis reductions in `ultralytics/utils/tal.py` with direct elementwise operations and `max` calls, reducing CPU and MPS overhead while preserving reported outputs.

### 📊 Key Changes

- `RotatedTaskAlignedAssigner.select_candidates_in_gts` now computes the two-coordinate dot products directly on separate tensors instead of creating a final size-2 dimension and reducing it.
- `TaskAlignedAssigner.select_highest_overlaps` replaces middle-axis `argmax` calls with `max(...).indices` and derives `fg_mask` and `target_gt_idx` together after duplicate assignments are resolved.
- Removes the subsequent `sum(-2)` operations because each anchor has at most one assigned ground-truth box at that stage.
- Reported benchmarks show substantial CPU speedups for both assignment methods, with smaller improvements on CUDA and little to no change for some MPS criterion runs.
- The PR reports identical outputs across its tested cases, including real inputs, multiple dtypes, ties, special values, non-contiguous tensors, and training runs.

### 🎯 Purpose & Impact

- Training and inference paths using task assignment perform less intermediate allocation and fewer reductions, particularly improving measured CPU performance and reducing CPU overhead on MPS.
- Assignment outputs, foreground masks, target indices, losses, and final weights were reported unchanged in the supplied validation runs.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the complete 9-line change in `ultralytics/utils/tal.py`. The algebraic rewrite is shape-compatible and removes the temporary coordinate dimension, but it may alter CUDA low-precision assignment masks because it drops the float accumulation used by the original reduction. One medium-risk compatibility issue remains.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/tal.py:418` This changes numerical behavior for CUDA fp16/bf16 inputs: the previous `sum(dim=-1)` uses a float accumulator for these dtypes, while these pointwise products and the final addition are performed in the tensor dtype. For anchors whose dot product is close to 0 or the edge norm, rounding can therefore flip the inclusive comparisons and change assignments. CUDA is explicitly unmeasured in the PR; preserve the previous accumulation precision for low-precision inputs or add CUDA fp16/bf16 coverage…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 1, the diff is unchanged and the CUDA low-precision numerical-compatibility concern remains unresolved; no new issue was introduced. The production OBB path currently passes float32 ground-truth boxes, but the changed assigner method itself has no dtype restriction and CUDA remains unvalidated.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/tal.py:418` Still open from review 1: for homogeneous `float16`/`bfloat16` inputs on CUDA, the old `sum(dim=-1)` uses float accumulation for reduced-precision values, while these pointwise products and the addition remain in the low-precision dtype. The inclusive comparisons on line 420 can therefore change for anchors near box edges; CUDA is explicitly unmeasured in this PR. PyTorch's accumulation types confirm CUDA half and bfloat16 use float accumulation for reductions.  Preserve the previous accumulati…
</details>

File: ultralytics/utils/tal.py [hunk 1]
```diff
@@ -410,12 +410,13 @@ def select_candidates_in_gts(self, xy_centers, gt_bboxes, mask_gt):
...
-        # (b, n_boxes, h*w, 2)
-        ap = xy_centers - a
+        # (b, n_boxes, h*w) per coordinate
+        apx = xy_centers[:, 0] - a[..., 0]
+        apy = xy_centers[:, 1] - a[..., 1]
         norm_ab = (ab * ab).sum(dim=-1)
         norm_ad = (ad * ad).sum(dim=-1)
-        ap_dot_ab = (ap * ab).sum(dim=-1)
-        ap_dot_ad = (ap * ad).sum(dim=-1)
+        ap_dot_ab = apx * ab[..., 0] + apy * ab[..., 1]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py]: 💡 **MEDIUM**: Still open from review 1/2: for homogeneous CUDA `float16`/`bfloat16` inputs, the previous `(ap * ab).sum(dim=-1)` performs the reduction with a float accumulator, while this expression performs the final `+` in the low-precision tensor dtype, rounding before the comparison. PyTorch's CUDA reduction implementation explicitly dispatches half and bfloat16 sums with float accumulation. Anchors near an edge can therefore produce a different inclusive mask and alter assignment targets; CUDA is still unmeasured in this PR. Preserve the previous accumulation behavior or add CUDA fp16/bf16 equivalence coverage before merging. ([github.com](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/ReduceSumProdKernel.cu?utm_source=openai))

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 2, the 9-line diff is unchanged. The shape-compatible rewrite and reported CPU/MPS behavior look sound, but the unresolved CUDA low-precision numerical compatibility risk remains. No other concrete issue was found.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/tal.py:418` Still open from review 1/2: for homogeneous CUDA `float16`/`bfloat16` inputs, the previous `(ap * ab).sum(dim=-1)` performs the reduction with a float accumulator, while this expression performs the final `+` in the low-precision tensor dtype, rounding before the comparison. PyTorch's CUDA reduction implementation explicitly dispatches half and bfloat16 sums with float accumulation. Anchors near an edge can therefore produce a different inclusive mask and alter assignment targets; CUDA is still…
</details>

File: ultralytics/utils/tal.py [hunk 2]
```diff
@@ -410,12 +408,13 @@ def select_candidates_in_gts(self, xy_centers, gt_bboxes, mask_gt):
...
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py]: 💡 **MEDIUM**: This still changes the CUDA low-precision behavior flagged in reviews 1–3. For homogeneous CUDA `float16`/`bfloat16` inputs, the old `(ap * ab).sum(dim=-1)` uses a float accumulation type, while this pointwise `+` is performed in the input dtype; values near 0 or `norm_ab` can therefore round differently and flip the inclusive comparisons on line 418, changing assignment masks. The production OBB path commonly supplies float32 ground-truth boxes, but this method has no dtype restriction and CUDA remains unvalidated; CPU/MPS differential tests do not establish CUDA equivalence. Preserve the previous accumulation semantics for reduced dtypes or add actual CUDA fp16/bf16 differential coverage before accepting the rewrite. ([github.com](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/AccumulateType.h?utm_source=openai))

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 3, the code diff is unchanged; the new description adds CPU/MPS and training-path evidence but does not resolve the previously identified CUDA low-precision compatibility risk. The `select_highest_overlaps` rewrite has no additional concrete issue found. One medium-risk numerical concern remains open.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/utils/tal.py:416` This still changes the CUDA low-precision behavior flagged in reviews 1–3. For homogeneous CUDA `float16`/`bfloat16` inputs, the old `(ap * ab).sum(dim=-1)` uses a float accumulation type, while this pointwise `+` is performed in the input dtype; values near 0 or `norm_ab` can therefore round differently and flip the inclusive comparisons on line 418, changing assignment masks. The production OBB path commonly supplies float32 ground-truth boxes, but this method has no dtype restriction and CUD…
</details>

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since review 4, the previously raised CUDA low-precision concern is addressed by the reported CUDA fp16/bf16 fuzz and training validation. The reduction rewrites preserve the relevant assignment behavior and no new concrete regression is visible in the current diff. LGTM.

**Comment by UltralyticsAssistant** [comment]
🎉 PR [#26234](https://github.com/ultralytics/ultralytics/pull/26234) has been merged—thank you, @raimbekovm!

> “Well done is better than well said.” — Benjamin Franklin

This optimization replaces small tensor reductions with direct elementwise operations and `max` calls, reducing CPU and MPS overhead while preserving reported assignment outputs, losses, and final weights. Excellent work!
